Fix type promotion for trace on CPU.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47305 Fix type promotion for trace on CPU.**

Fixes https://github.com/pytorch/pytorch/issues/47127.

Ideally this would just use diag and sum (as the CUDA implementation does), but that seems to have performance problems, which I'll link in the github PR.

Differential Revision: [D24729627](https://our.internmc.facebook.com/intern/diff/D24729627)

# BC-breaking message:

Previously, the cpu variant of trace did not properly type promote, but the cuda variant did.  Both variants now type promote correctly:

1.7:
```
>>> import torch
>>> x = torch.ones(5, 5, dtype=torch.uint8)
>>> x_cuda = x.cuda()
>>> x.trace().dtype
torch.uint8
>>> x_cuda.trace().dtype
torch.int64
```

1.8:
```
>>> import torch
>>> x = torch.ones(5, 5, dtype=torch.uint8)
>>> x_cuda = x.cuda()
>>> x.trace().dtype
torch.int64
>>> x_cuda.trace().dtype
torch.int64
```

gchanan committed:
Fix type promotion for trace on CPU.

Fixes https://github.com/pytorch/pytorch/issues/47127.

Ideally this would just use diag and sum (as the CUDA implementation does), but that seems to have performance problems, which I'll link in the github PR.

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -473,6 +474,38 @@ static Tensor& prod_out_impl(Tensor& result, const Tensor& self, IntArrayRef dim
   return result;
 }
 
+Tensor trace_cpu(const Tensor& self) {
```
nit: The best place to put this might be TriangularOps.cpp (trace_cuda is in TriangularOps.cu). Alternatively we can move trace_cuda out of TriangularOps.cu for consistency

gchanan replied:
the problem is you need the specialized promotion logic from here.  In theory you could split things up but this doesn't seem worth it given the promotion logic is really specialized for reductions.

gchanan commented:
Here's some benchmarking results for this vs the `diag().sum()` option:

This:
```
 1:
>>> t.timeit.timeit(setup='import torch; x=torch.ones(1, 1, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.2587152142077684
>>> t.timeit.timeit(setup='import torch; x=torch.ones(1, 1, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.240397651679814
>>> t.timeit.timeit(setup='import torch; x=torch.ones(1, 1, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.2281987024471164
>>> t.timeit.timeit(setup='import torch; x=torch.ones(1, 1, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.2414425648748875

300:
>>> t.timeit.timeit(setup='import torch; x=torch.ones(300, 300, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.571804345585406
>>> t.timeit.timeit(setup='import torch; x=torch.ones(300, 300, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.574404393322766
>>> t.timeit.timeit(setup='import torch; x=torch.ones(300, 300, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.676772426813841
>>> t.timeit.timeit(setup='import torch; x=torch.ones(300, 300, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.566413162276149
>>> t.timeit.timeit(setup='import torch; x=torch.ones(300, 300, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
2.5903642047196627

100000
>>> t.timeit.timeit(setup='import torch; x=torch.ones(100000, 100000, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000)
2.908080894500017
>>> t.timeit.timeit(setup='import torch; x=torch.ones(100000, 100000, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000)
3.2296120524406433
```

`diag().sum()`:
```
1
>>> t.timeit.timeit(setup='import torch; x=torch.ones(1, 1, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
8.96580315567553
>>> t.timeit.timeit(setup='import torch; x=torch.ones(1, 1, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
9.246467645280063
>>> t.timeit.timeit(setup='import torch; x=torch.ones(1, 1, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
9.167060194537044

300
>>> t.timeit.timeit(setup='import torch; x=torch.ones(300, 300, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
10.824241297319531
>>> t.timeit.timeit(setup='import torch; x=torch.ones(300, 300, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000000)
10.658952249214053

100000
>>> t.timeit.timeit(setup='import torch; x=torch.ones(100000, 100000, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000)
2.981750411912799
>>> t.timeit.timeit(setup='import torch; x=torch.ones(100000, 100000, dtype=torch.uint8)', stmt='torch.trace(x)', number=1000)
3.28736359719187
```

gchanan commented:
so the overhead is much higher, although for big sizes it doesn't matter.

zou3519 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -473,6 +474,38 @@ static Tensor& prod_out_impl(Tensor& result, const Tensor& self, IntArrayRef dim
+  });
+
+  return result;
+}
```
We're going to have to update test_trace:
https://github.com/pytorch/pytorch/blob/774b638eb64acb386f560603adb025e0567d5215/test/test_torch.py#L5996-L6029

zou3519 commented:
Should we mark this as BC-breaking? In older versions of PyTorch (<1.5), we did not do any type promotions for trace

gchanan commented:
ya I'll mark it BC-breaking.

gchanan committed:
Update on "Fix type promotion for trace on CPU."


Fixes https://github.com/pytorch/pytorch/issues/47127.

Ideally this would just use diag and sum (as the CUDA implementation does), but that seems to have performance problems, which I'll link in the github PR.

[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -5993,15 +5993,12 @@ def test_diagonal_multidim(self, device, dtype):
+    def test_trace(self, device, dtype):
         def test(shape):
             tensor = make_tensor(shape, device, dtype, low=-9, high=9)
             diag = tensor.diag()
```
The diag = tensor.diag() line is bogus, can you remove it please? (Sorry, I was the one who added it but I am only noticing it now)

zou3519 commented on test/test_torch.py:
```diff
@@ -5993,15 +5993,12 @@ def test_diagonal_multidim(self, device, dtype):
         self.assertEqual(expected, result)
 
-    def _test_trace(self, device, dtype, legacy):
+    @dtypes(*torch.testing.get_all_dtypes(include_complex=False, include_bool=False, include_half=False, include_bfloat16=False))
```
We've dropped testing for half on CUDA, is there a good way around this?

The dtypesIfCUDA and dtypesifCPU decorators look like they handle this (albeit very verbosely):
```
@dtypesIfCPU(*torch.testing.get_all_dtypes(include_complex=False, include_bool=False, include_half=False, include_bfloat16=False))
@dtypesIfCUDA(*torch.testing.get_all_dtypes(include_complex=False, include_bool=False, include_bfloat16=False))
```

gchanan committed:
Update on "Fix type promotion for trace on CPU."


Fixes https://github.com/pytorch/pytorch/issues/47127.

Ideally this would just use diag and sum (as the CUDA implementation does), but that seems to have performance problems, which I'll link in the github PR.

[ghstack-poisoned]

gchanan committed:
Update on "Fix type promotion for trace on CPU."


Fixes https://github.com/pytorch/pytorch/issues/47127.

Ideally this would just use diag and sum (as the CUDA implementation does), but that seems to have performance problems, which I'll link in the github PR.

[ghstack-poisoned]

gchanan committed:
Update on "Fix type promotion for trace on CPU."


Fixes https://github.com/pytorch/pytorch/issues/47127.

Ideally this would just use diag and sum (as the CUDA implementation does), but that seems to have performance problems, which I'll link in the github PR.

Differential Revision: [D24729627](https://our.internmc.facebook.com/intern/diff/D24729627)

gchanan committed:
Update on "Fix type promotion for trace on CPU."


Fixes https://github.com/pytorch/pytorch/issues/47127.

Ideally this would just use diag and sum (as the CUDA implementation does), but that seems to have performance problems, which I'll link in the github PR.

Differential Revision: [D24729627](https://our.internmc.facebook.com/intern/diff/D24729627)

gchanan committed:
Update on "Fix type promotion for trace on CPU."


Fixes https://github.com/pytorch/pytorch/issues/47127.

Ideally this would just use diag and sum (as the CUDA implementation does), but that seems to have performance problems, which I'll link in the github PR.

Differential Revision: [D24729627](https://our.internmc.facebook.com/intern/diff/D24729627)
